src/server/main_service.cc

```diff
@@ -1565,52 +1564,66 @@ void Service::EvalInternal(const EvalArgs& eval_args, Interpreter* interpreter,
 
   DCHECK(!cntx->conn_state.script_info);  // we should not call eval from the script.
 
+  optional<ShardId> sid;
+
   // TODO: to determine whether the script is RO by scanning all "redis.p?call" calls
   // and checking whether all invocations consist of RO commands.
   // we can do it once during script insertion into script mgr.
   auto& sinfo = cntx->conn_state.script_info;
   sinfo = make_unique<ConnectionState::ScriptInfo>();
   for (size_t i = 0; i < eval_args.keys.size(); ++i) {
-    sinfo->keys.insert(KeyLockArgs::GetLockKey(ArgS(eval_args.keys, i)));
+    string_view key = KeyLockArgs::GetLockKey(ArgS(eval_args.keys, i));
+    sinfo->keys.insert(key);
+
+    ShardId cur_sid = Shard(key, shard_count());
+    if (i == 0) {
+      sid = cur_sid;
+    }
+    if (sid.has_value() && *sid != cur_sid) {
+      sid = nullopt;
+    }
   }
+
   sinfo->async_cmds_heap_limit = absl::GetFlag(FLAGS_multi_eval_squash_buffer);
   Transaction* tx = cntx->transaction;
   CHECK(tx != nullptr);
 
-  optional<bool> scheduled = StartMultiEval(cntx->db_index(), eval_args.keys, *params, cntx);
-  if (!scheduled) {
-    return;
-  }
-
   interpreter->SetGlobalArray("KEYS", eval_args.keys);
   interpreter->SetGlobalArray("ARGV", eval_args.args);
+  absl::Cleanup clean = [interpreter]() { interpreter->ResetStack(); };
 
   Interpreter::RunResult result;
-  optional<ShardId> sid = GetRemoteShardToRunAt(*tx);
-  if (tx->GetMultiMode() != Transaction::NON_ATOMIC && sid.has_value()) {
+
+  if (CanRunSingleShardMulti(sid, *params, tx)) {
     // If script runs on a single shard, we run it remotely to save hops.
-    interpreter->SetRedisFunc([cntx, this](Interpreter::CallArgs args) {
+    interpreter->SetRedisFunc([cntx, tx, this](Interpreter::CallArgs args) {
       // Disable squashing, as we're using the squashing mechanism to run remotely.
       args.async = false;
       CallFromScript(cntx, args);
     });
 
-    boost::intrusive_ptr<Transaction> stub_tx = new Transaction{cntx->transaction};
-    cntx->transaction = stub_tx.get();
-    tx->PrepareSquashedMultiHop(registry_.Find("EVAL"), [&](ShardId id) { return id == *sid; });
-    tx->ScheduleSingleHop([&](Transaction* tx, EngineShard*) {
+    tx->RunSingleShardMulti(cntx->db_index(), eval_args.keys, [&]() {
       ++ServerState::tlocal()->stats.eval_shardlocal_coordination_cnt;
+      boost::intrusive_ptr<Transaction> stub_tx = new Transaction{tx};
```

Comment:
i would pull out the code that does not have to be in the callback. i.e. lines 1606-1608